pkg/storegateway/gateway_prepare_shutdown_http_test.go

```diff
@@ -0,0 +1,163 @@
+	require.True(t, g.ringLifecycler.ShouldKeepInstanceInTheRingOnShutdown())
+
+	// after GET is invoked, the expected result is "unset"
+	response1 := httptest.NewRecorder()
+	g.PrepareShutdownHandler(response1, httptest.NewRequest("GET", "/store-gateway/prepare-shutdown", nil))
+	require.Equal(t, "unset", response1.Body.String())
+	require.Equal(t, 200, response1.Code)
+	exists, err = shutdownmarker.Exists(shutdownMarkerPath)
+	require.Nil(t, err)
+	require.False(t, exists)
+
+	// after POST is invoked, it is required that cortex_storegateway_prepare_shutdown_requested gets incremented
+	// and that there exists a shutdown marker
+	response2 := httptest.NewRecorder()
+	g.PrepareShutdownHandler(response2, httptest.NewRequest("POST", "/store-gateway/prepare-shutdown", nil))
+	require.Equal(t, 204, response2.Code)
+
+	require.NoError(t, testutil.GatherAndCompare(reg, bytes.NewBufferString(`
+					# HELP cortex_storegateway_prepare_shutdown_requested If the store-gateway has been requested to prepare for shutdown via endpoint or marker file.
+					# TYPE cortex_storegateway_prepare_shutdown_requested gauge
+					cortex_storegateway_prepare_shutdown_requested 1
+				`), "cortex_storegateway_prepare_shutdown_requested"))
+
+	exists, err = shutdownmarker.Exists(shutdownMarkerPath)
+	require.Nil(t, err)
+	require.True(t, exists)
+	// ensure that unregistering is required
+	require.False(t, g.ringLifecycler.ShouldKeepInstanceInTheRingOnShutdown())
+
+	// after GET is invoked, the expected result is now "set"
+	response3 := httptest.NewRecorder()
+	g.PrepareShutdownHandler(response3, httptest.NewRequest("GET", "/store-gateway/prepare-shutdown", nil))
+	require.Equal(t, "set", response3.Body.String())
+	require.Equal(t, 200, response3.Code)
+
+	// after DELETE is invoked, the effects of POST get deleted
+	response4 := httptest.NewRecorder()
+	g.PrepareShutdownHandler(response4, httptest.NewRequest("DELETE", "/store-gateway/prepare-shutdown", nil))
+	require.Equal(t, 204, response4.Code)
+
+	require.NoError(t, testutil.GatherAndCompare(reg, bytes.NewBufferString(`
+					# HELP cortex_storegateway_prepare_shutdown_requested If the store-gateway has been requested to prepare for shutdown via endpoint or marker file.
+					# TYPE cortex_storegateway_prepare_shutdown_requested gauge
+					cortex_storegateway_prepare_shutdown_requested 0
+				`), "cortex_storegateway_prepare_shutdown_requested"))
+	exists, err = shutdownmarker.Exists(shutdownMarkerPath)
+	require.Nil(t, err)
+	require.False(t, exists)
+	// ensure that unregistering is NOT required
+	require.True(t, g.ringLifecycler.ShouldKeepInstanceInTheRingOnShutdown())
+
+	// after POST is invoked, and store-gateway is stopped, it is required that it gets removed from the ring
+	response5 := httptest.NewRecorder()
+	g.PrepareShutdownHandler(response5, httptest.NewRequest("POST", "/store-gateway/prepare-shutdown", nil))
+	require.Equal(t, 204, response5.Code)
+
+	// Stop the store-gateway
```

Comment:
Not sure these assertions add much value because you're asserting on the lifecycler itself. A better assertion would be checking the actual state of the ring from the "ring store" created in `createStoreGateway()`. You can get the actual `RingDesc` calling `ringStore.Get()` and then assert there are no instances registered in the ring.